src/dbnode/encoding/proto/common.go

```diff
@@ -0,0 +1,268 @@
+	// of the fields for multiple types to save memory, but its ok for now.
+	fieldNum  int
+	fieldType customFieldType
+
+	// Float state.
+	floatXORState m3tsz.XOREncoder
+
+	// Bytes State.
+	bytesFieldDict         []encoderBytesFieldDictState
+	iteratorBytesFieldDict [][]byte
+
+	// Int state.
+	prevIntBits       uint64
+	intSigBitsTracker m3tsz.IntSigBitsTracker
+}
+
+type encoderBytesFieldDictState struct {
+	// We store the hash so we can perform fast equality checks, and
+	// we store the startPos + length so that when we have a value
+	// that matches a hash, we can be certain its not a hash collision
+	// by comparing the bytes against those we already wrote into the
+	// stream.
+	hash     uint64
+	startPos int
+	length   int
+}
+
+func newCustomFieldState(fieldNum int, fieldType customFieldType) customFieldState {
+	return customFieldState{fieldNum: fieldNum, fieldType: fieldType}
+}
+
+// TODO(rartoul): Improve this function to be less naive and actually explore nested messages
+// for fields that we can use our custom compression on: https://github.com/m3db/m3/issues/1471
+func customFields(s []customFieldState, schema *desc.MessageDescriptor) []customFieldState {
+	numCustomFields := numCustomFields(schema)
+	if cap(s) >= numCustomFields {
+		s = s[:0]
+	} else {
+		s = make([]customFieldState, 0, numCustomFields)
+	}
+
+	fields := schema.GetFields()
+	for _, field := range fields {
+		customFieldType, ok := isCustomField(field.GetType(), field.IsRepeated())
+		if !ok {
+			continue
+		}
+
+		fieldState := newCustomFieldState(int(field.GetNumber()), customFieldType)
+		s = append(s, fieldState)
+	}
+
+	// Should already be sorted by fieldNum, but do it again just to be sure.
+	sort.Slice(s, func(a, b int) bool {
+		return s[a].fieldNum < s[b].fieldNum
+	})
+
+	return s
+}
+
+func isCustomFloatEncodedField(t customFieldType) bool {
+	return t == cFloat64 || t == cFloat32
+}
+
+func isCustomIntEncodedField(t customFieldType) bool {
+	return t == cSignedInt64 ||
+		t == cUnsignedInt64 ||
+		t == cSignedInt32 ||
+		t == cUnsignedInt32
+}
+
+func isUnsignedInt(t customFieldType) bool {
+	return t == cUnsignedInt64 || t == cUnsignedInt32
+}
+
+func numCustomFields(schema *desc.MessageDescriptor) int {
+	var (
+		fields          = schema.GetFields()
+		numCustomFields = 0
+	)
+
+	for _, field := range fields {
+		if _, ok := isCustomField(field.GetType(), field.IsRepeated()); ok {
+			numCustomFields++
+		}
+	}
+
+	return numCustomFields
+}
+
+func isCustomField(fieldType dpb.FieldDescriptorProto_Type, isRepeated bool) (customFieldType, bool) {
+	if isRepeated {
+		return -1, false
+	}
+
+	customFieldType, ok := mapProtoTypeToCustomFieldType[fieldType]
+	return customFieldType, ok
+}
+
+func resetCustomFields(fields []customFieldState, schema *desc.MessageDescriptor) []customFieldState {
+	if cap(fields) <= maxTSZFieldsCapacityRetain {
+		return customFields(fields, schema)
+	}
+	return customFields(nil, schema)
```

Comment:
Do we want to make maxTSZFieldsCapacityRetain configurable? I imagine in use cases where you have more than 24 fields in a proto schema, you'd always have to realloc the fields?

Reply:
Honestly, why not for now just get rid of `maxTSZFieldsCapacityRetain`?  Seems like a premature optimization perhaps.  

If it becomes a problem with tons of people using a lot of fields and it starts to show up in profiles we can add this at that time.  As far as I can see pretty much most workloads are going to have basically the exact same number of fields being used (since schema doesn't change between any one encoder in a namespace currently) so this will always be the same and it's always worth reusing the fields slice.

Reply:
ok fine to do it for now